src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -3107,5 +3107,76 @@ .locals init (char V_0)
   IL_006d:  ret
 }");
         }
+
+        [Fact]
+        public void OutputType_01()
+        {
+            var source = @"using System;
+class C
+{
+    static void Main()
+    {
+        // 1. If P is a type pattern, the narrowed type is the type of the type pattern's type.
+        object o = 1;
+        { if (o is int and var i) M(i); } // System.Int32
+
+        // 2. If P is a declaration pattern, the narrowed type is the type of the declaration pattern's type.
+        o = 1L;
+        { if (o is long q and var i) M(i); } // System.Int64
+
+        // 2. If P is a recursive pattern that gives an explicit type, the narrowed type is that type.
+        o = 1UL;
+        { if (o is ulong {} and var i) M(i); } // System.UInt64
+
+        // 4. If P is a constant pattern where the constant is not the null constant and where the expression has no constant expression conversion to the input type, the narrowed type is the type of the constant.
```

Comment:
Is there any need to test the case where there is a constant expression conversion. e.g. `long x = 0; if (x is 42 and var i) { M(i); } // expect System.Int64`

Reply:
Sure (that's case 8)

---
In reply to: [389164346](https://github.com/dotnet/roslyn/pull/42207#discussion_r389164346) [](ancestors = 389164346)